config/cmake/version_describe.cmake

```diff
@@ -0,0 +1,45 @@
+#
+# Copyright (c) 2017-2019, Battelle Memorial Institute; Lawrence Livermore National
+# Security, LLC; Alliance for Sustainable Energy, LLC.  See the top-level NOTICE for additional details. All rights reserved. 
+# SPDX-License-Identifier: BSD-3-Clause
+#
+
+find_package(Git)
+
+# this function is to get a nicely formatted build string for later reference in a human readable format that has some utility for automation.  
+# the resulting format would be empty if the build matches a tag exactly, and otherwise <branch_name>-<git hash>[-dirty]  if dirty is present there are uncommitted changes in the repo
+# called like git_version_describe(${PROJECT_SOURCE_DIR} result)  then result will contain the description string
+
+function(git_version_describe source_path result)
+  message(STATUS "Source ${source_path}")
+  if (GIT_FOUND)
+	execute_process(COMMAND ${GIT_EXECUTABLE} -C ${source_path} --work-tree ${source_path} describe --tags OUTPUT_VARIABLE TAG_DESCRIPTION OUTPUT_STRIP_TRAILING_WHITESPACE)
+	string(LENGTH ${TAG_DESCRIPTION} tag_length)
+	if (tag_length GREATER 16)
+	    string(FIND ${TAG_DESCRIPTION} "-" last_dash_loc REVERSE)
+		string(SUBSTRING ${TAG_DESCRIPTION} ${last_dash_loc} -1 hash_string)
+		execute_process(COMMAND ${GIT_EXECUTABLE} -C ${source_path} --work-tree ${source_path} describe --all --tags --dirty OUTPUT_VARIABLE TAG_DESCRIPTION2 OUTPUT_STRIP_TRAILING_WHITESPACE)
+		string(FIND ${TAG_DESCRIPTION2} "-dirty" dirty_loc)
+		if (dirty_loc GREATER 0)
+			string(SUBSTRING ${TAG_DESCRIPTION2} 0 ${dirty_loc} tag_desc)
+			string(APPEND tag_desc ${hash_string} "-dirty")
+		else()
+			set(tag_desc ${TAG_DESCRIPTION2})
+			string(APPEND tag_desc ${hash_string})
+		endif()
+		
+		string(FIND ${tag_desc} "heads/" hloc)
```

Comment:
Should this get done before appending "-dirty" to the tag description?

Reply:
I don't think it matters as heads would be at the beginning, and I can't say I am particularly concerned with the performance of the string processing algorithms in cmake